File: src/RqCalc.Wpf/Models/CharacterStatEditModel.cs
using Framework.Reactive;
using Framework.Reactive.ObservableRecurse;

using RqCalc.Domain;
using RqCalc.Model;

namespace RqCalc.Wpf.Models;

public class CharacterStatEditModel : NotifyModelBase
{
    private readonly ApplicationSettings settings;
    private readonly CharacterChangeModel rootModel;

    public CharacterStatEditModel(
        ApplicationSettings settings,
        CharacterChangeModel rootModel)
    {
        this.settings = settings;
        this.rootModel = rootModel;

        this.SubscribeExplicit(rule => rule.Subscribe(model => model.DisplayValue, () =>
        {
            this.MultiplicityValue = this.DisplayValue / this.Multiplicity;
            this.MultiplicityValueVisible = this.MultiplicityValue != 0;
            this.ProgressValue = this.DisplayValue % this.Multiplicity;
        }));
    }

    public void SetMaxStat()
    {
        var limit = this.settings.MaxStatCount - this.EditValue;

        if (limit > 0)
        {
            this.EditValue += Math.Min(this.rootModel.FreeStats, limit);
        }
    }

    public void TryIncreaseStat()
    {
        var limit = this.settings.MaxStatCount - this.EditValue;

        if (limit > 0 && this.rootModel.FreeStats > 0)
        {
            this.EditValue++;
        }
    }

    public void TryDecreaseStat()
    {
        if (this.EditValue > 1)
        {
            this.EditValue--;
        }
    }

    public int EditValue
    {
        get => this.GetValue(v => v.EditValue);
        set => this.SetValue(v => v.EditValue, value);
    }

    public int DisplayValue
    {
        get => this.GetValue(v => v.DisplayValue);
        set => this.SetValue(v => v.DisplayValue, value);
    }

    public int Multiplicity => this.settings.LevelMultiplicity;

    public int MultiplicityValue
    {
        get => this.GetValue(v => v.MultiplicityValue);
        private set => this.SetValue(v => v.MultiplicityValue, value);
    }

    public int ProgressValue
    {
        get => this.GetValue(v => v.ProgressValue);
        private set => this.SetValue(v => v.ProgressValue, value);
    }

    public bool MultiplicityValueVisible
    {
        get => this.GetValue(v => v.MultiplicityValueVisible);
        private set => this.SetValue(v => v.MultiplicityValueVisible, value);
    }


    public IStat Stat
    {
        get => this.GetValue(v => v.Stat);
        set => this.SetValue(v => v.Stat, value);
    }
}

File: src/RqCalc.WebApp/_Scripts/Models/CharacterStatEditModel.js
function CharacterStatEditModel(context, rootModel, displayModel) {
    
    var thisModel = this;

    thisModel.displayModel = displayModel;
    thisModel.editValue = ko.observable();
    thisModel.displayValue = ko.observable();

    thisModel.multiplicityValue = ko.observable();
    thisModel.progressPercentValue = ko.observable();

    thisModel.multiplicity = context.settings.levelMultiplicity;
    thisModel.multiplicityValueVisible = ko.observable();
    thisModel.stat = ko.observable(displayModel.stat());
    
    thisModel.displayValue.subscribe(function (_) {

        thisModel.multiplicityValue(Math.floor(thisModel.displayValue() / thisModel.multiplicity));
        thisModel.multiplicityValueVisible(thisModel.multiplicityValue() !== 0);

        thisModel.progressPercentValue((((thisModel.displayValue() % thisModel.multiplicity) / thisModel.multiplicity) * 100).round(0));
    });

    displayModel.value.subscribe(function (newValue) {
        thisModel.displayValue(newValue);
    });

    function getLimit() {
        return context.settings.maxStatCount - thisModel.editValue();
    };

    thisModel.maximize = function() {

        var limit = getLimit();

        if (limit > 0 && rootModel.freeStats() > 0) {
            thisModel.editValue(thisModel.editValue() + Math.min(rootModel.freeStats(), limit));
        }
    };

    thisModel.increment = function () {

        var limit = getLimit();
        
         if (limit > 0 && rootModel.freeStats() > 0) {
             thisModel.editValue(thisModel.editValue() + 1);
         }
    };

    thisModel.decrement = function () {

        if (thisModel.editValue() > 1) {
            thisModel.editValue(thisModel.editValue() - 1);
        }
    };

    thisModel.description = ko.computed (function () {

        var editValue = thisModel.editValue();

        var delta = thisModel.displayValue() - editValue;

        if (delta === 0) {
            return editValue;
        } else {
            return editValue + " + " + delta;
        }

    }, thisModel);
}